aws/rust-runtime/aws-auth/src/provider/lazy_caching.rs

```diff
@@ -0,0 +1,463 @@
+        .expiry()
+        .expect("refresh sets expiry if not given");
+    now > expiration
+}
+
+#[derive(Clone)]
+struct Inner<T: TimeProvider> {
+    time: T,
+    cache: Cache,
+    refresh: Arc<dyn AsyncProvideCredentials>,
+    refresh_timeout: Duration,
+    default_credential_expiration: Duration,
+}
+
+impl<T: TimeProvider> Inner<T> {
+    async fn refresh(&self) -> CredentialsResult {
+        let time = self.time.clone();
+        let default_credential_expiration = self.default_credential_expiration;
+        let future = self.refresh.provide_credentials();
+        self.cache
+            .refresh(|| async move {
+                let credentials = future.await?;
+                // If the credentials don't have an expiration time, then create a default one
+                let credentials = if credentials.expiry().is_none() {
+                    Credentials::new(
+                        credentials.access_key_id(),
+                        credentials.secret_access_key(),
+                        credentials.session_token().map(|s| s.to_string()),
+                        Some(time.now() + default_credential_expiration),
+                        "lazy_caching_default_credential_expiration",
+                    )
+                } else {
+                    credentials
+                };
+                Ok(credentials)
+            })
+            .await
+    }
+
+    async fn needs_refresh(&self, now: SystemTime) -> bool {
+        if let Some(credentials) = self.cache.get().await {
+            if expired(&credentials, now) {
+                self.cache.clear_if_expired(now).await
+            } else {
+                false
+            }
+        } else {
+            true
+        }
+    }
+
+    async fn cached(&self) -> Credentials {
+        self.cache
+            .get()
+            .await
+            .expect("refresh requirement checked in advance")
+    }
+}
+
+struct Provider<T: TimeProvider> {
+    inner: Inner<T>,
+}
+
+impl<T: TimeProvider> Provider<T> {
+    fn new(
+        time: T,
+        refresh: Arc<dyn AsyncProvideCredentials>,
+        refresh_timeout: Duration,
+        default_credential_expiration: Duration,
+    ) -> Self {
+        Provider {
+            inner: Inner {
+                time,
+                cache: Cache::new(),
+                refresh,
+                refresh_timeout,
+                default_credential_expiration,
+            },
+        }
+    }
+
+    fn provide_credentials(&self) -> BoxFuture<CredentialsResult> {
+        let inner = self.inner.clone();
+        Box::pin(async move {
+            let now = inner.time.now();
+            if inner.needs_refresh(now).await {
+                let span = trace_span!("lazy_refresh_credentials");
+                let _enter = span.enter();
+                inner.refresh().await
+            } else {
+                Ok(inner.cached().await)
+            }
+        })
+    }
+}
+
+#[derive(Clone)]
+struct Cache {
```

Comment:
I think we should probably extract this into its own module and add loom tests. The read / reload behavior is pretty subtle and it would hard to track down a potential race condition.

Reply:
Moved `Cache` into its own module but ended up reworking some of the retrieve/refresh logic since I found a bug in it. Attempted to write a loom test for this, but loom is missing mock implementations for Tokio RwLock and OnceCell, so I opted for running a multi-threaded test thousands of times.